### PR TITLE
Always use `trainer.call_hook`

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

Always try to use `trainer.call_hook` to call hooks as we need it to set the `_current_fx_name` attribute in the `LightningModule` which is required for logging.

Since `call_hook` also calls any trainer and accelerator methods with the same name, we need to skip them for `setup` and `teardown`.

A test is added which tries to `self.log()` in all `LightningModule` and `Callback` hooks to assert that `MisconfigurationException` is raised.

Part of #7898 

## Follow-ups

- The new test could assert the final logged values.
- We should refactor `call_hook`. The current pattern of always calling the accelerator and trainer hooks of the same name is dangerous. Also, we want the users to interact with it after loop customization. #8506 

### Does your PR introduce any breaking changes ? If yes, please list them.

None intended

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [n/a] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)
- [x] Did you list all the breaking changes introduced by this pull request?

## PR review
 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

### CODE REVIEW

File: pytorch_lightning/trainer/data_loading.py [hunk 1]
```diff
@@ -463,14 +463,15 @@ def reset_train_val_dataloaders(self, model) -> None:
...
         """Handles downloading data in the GPU or TPU case.
 
         Returns:
             The dataloader
         """
-        self.call_hook(f"on_{stage}_dataloader")
-        dataloader = getattr(model, f'{stage}_dataloader')()
+        hook = f"{stage}_dataloader"
+        self.call_hook("on_" + hook)
+        dataloader = self.call_hook(hook)
```
Comment (tchaton) [inline pytorch_lightning/trainer/data_loading.py:474]: Oh I see, this would be triggered only on the Trainer and not model.

File: pytorch_lightning/trainer/trainer.py [hunk 2]
```diff
@@ -1200,9 +1191,11 @@ def call_hook(self, hook_name: str, *args, **kwargs) -> Any:
         with self.profiler.profile(hook_name):
 
             # first call trainer hook
-            if hasattr(self, hook_name):
+            if hook_name not in ("setup", ) and hasattr(self, hook_name):
```
Comment (tchaton) [inline pytorch_lightning/trainer/trainer.py:1194]: Ideally, connectors should also be called using `call_hook`.

Comment (carmocca) [reply pytorch_lightning/trainer/trainer.py:1197]: Opened #8506 to discuss

File: pytorch_lightning/core/lightning.py [hunk 3]
```diff
@@ -405,8 +405,18 @@ def log(
         on_epoch = self.__auto_choose_log_on_epoch(on_epoch)
 
         results = self.trainer._results
-        assert results is not None
-        assert self._current_fx_name is not None
+        if results is None:
+            caller = inspect.stack()[1][3]
```
Comment (awaelchli) [inline pytorch_lightning/core/lightning.py:409]: based on my simple testing here I believe this will not return the right answer if self.log is called in a function nested under the `current_fx_name` function. Will this not be misleading?

File: tests/trainer/connectors/test_callback_connector.py [hunk 4]
```diff
@@ -89,10 +87,8 @@ def test_attach_model_callbacks():
     """Test that the callbacks defined in the model and through Trainer get merged correctly."""
 
     def assert_composition(trainer_callbacks, model_callbacks, expected):
-        model = Mock()
-        model.configure_callbacks.return_value = model_callbacks
         trainer = Trainer(checkpoint_callback=False, progress_bar_refresh_rate=0, callbacks=trainer_callbacks)
-        trainer.model = model
+        trainer.call_hook = Mock(return_value=model_callbacks)
```
Comment (awaelchli) [inline tests/trainer/connectors/test_callback_connector.py:91]: why can we no longer mock the method on the module?

File: pytorch_lightning/trainer/trainer.py [hunk 5]
```diff
@@ -1231,38 +1221,40 @@ def _call_teardown_hook(self) -> None:
...
 
             # next call hook in lightningModule
             output = None
-            model_ref = self.lightning_module
-            if is_overridden(hook_name, model_ref):
-                hook_fx = getattr(model_ref, hook_name)
-                output = hook_fx(*args, **kwargs)
+            model_fx = getattr(pl_module, hook_name, None)
+            if callable(model_fx):
+                output = model_fx(*args, **kwargs)
```
Comment (awaelchli) [inline pytorch_lightning/trainer/trainer.py:1244]: what are the implications? is it safe to call the default implementation? there must have been a reason in the past why we had is_overridden.

File: tests/trainer/connectors/test_callback_connector.py (hunk 4 above)
Comment (carmocca) [reply tests/trainer/connectors/test_callback_connector.py:91]: Because the previous implementation of `_attach_model_callbacks` did:

`self.trainer.model.configure_callbacks()`

but now it is

`self.trainer.call_hook("configure_callbacks")`


which means we can just mock the call hook function

Comment (carmocca) [reply tests/trainer/connectors/test_callback_connector.py:91]: Or are you asking why this changed? It was failing in the original implementation of this PR but seems like it's not anymore.

Do you prefer I undo these changes?

File: pytorch_lightning/trainer/trainer.py (hunk 5 above)
Comment (carmocca) [reply pytorch_lightning/trainer/trainer.py:1244]: Keeping the `is_overridden` check, so it doesn't call the base implementation

File: tests/trainer/connectors/test_callback_connector.py (hunk 4 above)
Comment (awaelchli) [reply tests/trainer/connectors/test_callback_connector.py]: is call_hook incompatible with mock?

if not, then mocking the call_hook here makes not much sense as this would result in it returning the model callbacks always, for all hook calls.

File: pytorch_lightning/core/lightning.py (hunk 3 above)
Comment (awaelchli) [reply pytorch_lightning/core/lightning.py:409]: does not matter

Comment (carmocca) [reply pytorch_lightning/core/lightning.py:409]: Actually, let me undo this because the exception prints the stacktrace already

I believe I did this because it made it easier to debug issues during the PR's implenentation

File: tests/trainer/connectors/test_callback_connector.py (hunk 4 above)
Comment (carmocca) [reply tests/trainer/connectors/test_callback_connector.py:88]: `call_hook` is compatible with mock.

> this would result in it returning the model callbacks always, for all hook calls.

True, but note that these are unit tests and just test the return value of a method in the connector.

**Review by awaelchli — APPROVED** [review]

File: tests/trainer/connectors/test_callback_connector.py (hunk 4 above)
Comment (carmocca) [reply tests/trainer/connectors/test_callback_connector.py:88]: Reverted these changes to keep the original pattern of the test. However, we cannot use a mock for the model anymore because of the `is_overridden` check in `call_hook`

**Review by Borda — COMMENTED** [review]
this will disable usage tracing in IDE, right?

**Comment by carmocca** [comment]
> this will disable usage tracing in IDE, right?

Yes, this was already the case. Most hooks are already using `call_hook`

File: pytorch_lightning/trainer/trainer.py (hunk 5 above)
Comment (carmocca) [reply pytorch_lightning/trainer/trainer.py]: Actually, just remembered why I need this.

The following tests fail with the `is_overridden` check:

```python
FAILED tests/callbacks/test_finetuning_callback.py::test_finetuning_callback
FAILED tests/callbacks/test_stochastic_weight_avg.py::test_swa_warns - ValueE...
FAILED tests/trainer/test_dataloaders.py::test_request_dataloader - Attribute...
FAILED tests/trainer/test_states.py::test_trainer_fn_while_running[Single-Step]
FAILED tests/trainer/test_trainer.py::test_model_in_correct_mode_during_stages[None-1]
FAILED tests/trainer/test_trainer.py::test_model_in_correct_mode_during_stages[ddp_cpu-2]
```

This is because the `on_*_model_{train,eval}` hooks do provide a default implementation so we need to call them even if they are not overridden.

Related to https://github.com/PyTorchLightning/pytorch-lightning/discussions/8760

**Review by awaelchli — APPROVED** [review]

**Review by justusschock — APPROVED** [review]

File: pytorch_lightning/trainer/trainer.py [hunk 6]
```diff
@@ -1184,8 +1177,7 @@ def _call_setup_hook(self) -> None:
 
         if self.datamodule is not None:
             self.datamodule.setup(stage=fn)
-        self.setup(stage=fn)
-        self.lightning_module.setup(stage=fn)
+        self.call_hook("setup", stage=fn)
```
Comment (tchaton) [inline pytorch_lightning/trainer/trainer.py:1180]: Could this change the current workflow ? I believe not, but just to raise attention.

**Review by tchaton — APPROVED** [review]
LGTM ! Great work !

File: pytorch_lightning/trainer/trainer.py (hunk 6 above)
Comment (carmocca) [reply pytorch_lightning/trainer/trainer.py:1180]: It should be good, the order is the same.
